fontbe/src/features/kern.rs

```diff
@@ -359,31 +390,843 @@ impl Work<Context, AnyWorkId, Error> for KerningGatherWork {
+        let lookups = self.finalize_kerning(&fragments, &ast.ast, &glyph_map, glyphs_and_gids)?;
+        context.fea_rs_kerns.set(lookups);
+        Ok(())
+    }
+}
+
+impl KerningGatherWork {
+    //. take the kerning fragments and generate the kerning lookups.
+    ///
+    // This includes much of the logic from the ufo2ft KernFeatureWriter
+    fn finalize_kerning(
+        &self,
+        fragments: &[&KernFragment],
+        ast: &ParseTree,
+        glyph_map: &GlyphMap,
+        glyphs: Vec<(Arc<Glyph>, GlyphId)>,
+    ) -> Result<FeaRsKerns, Error> {
+        // ignore diagnostics, they'll get logged during actual GSUB compilation
+        let (compilation, _) = fea_rs::compile::compile::<NopVariationInfo, NopFeatureProvider>(
+            ast,
+            glyph_map,
+            None,
+            None,
+            Opts::new().compile_gpos(false),
+        )
+        .map_err(|err| {
+            Error::FeaCompileError(fea_rs::compile::error::CompilerError::CompilationFail(err))
+        })?;
+
+        let gsub = compilation
+            .gsub
+            .as_ref()
+            .map(write_fonts::dump_table)
+            .transpose()
+            .expect("if this doesn't compile we will already panic when we try to add it to the context");
+        let gsub = gsub
+            .as_ref()
+            .map(|data| write_fonts::read::tables::gsub::Gsub::read(data.as_slice().into()))
+            .transpose()?;
+
+        let gdef = compilation.gdef_classes;
+
+        // serves as a standin for cmap in the fonttools impl
+
+        let pairs = fragments
+            .iter()
+            .flat_map(|frag| frag.kerns.iter())
+            .collect::<Vec<_>>();
+
+        let known_scripts = guess_font_scripts(ast, &glyphs);
+        let mark_glyphs = glyphs
+            .iter()
+            .filter_map(|(glyph, gid)| {
+                let is_mark = gdef
+                    .as_ref()
+                    .map(|gdef| gdef.get(gid) == Some(&GlyphClassDef::Mark))
+                    .unwrap_or(false);
+                is_mark.then(|| {
+                    let spacing = if glyph
+                        .sources()
+                        .values()
+                        .all(|instance| instance.width != 0.0)
+                    {
+                        MarkSpacing::Spacing
+                    } else {
+                        MarkSpacing::NonSpacing
+                    };
+                    (*gid, spacing)
+                })
+            })
+            .collect();
+        let split_ctx = KernSplitContext::new(&glyphs, &known_scripts, gsub, mark_glyphs)?;
+
+        let lookups = split_ctx.make_lookups(&pairs);
+        let (lookups, features) = self.assign_lookups_to_scripts(lookups, ast, KERN);
+        Ok(FeaRsKerns { lookups, features })
+    }
+
+    /// returns a vec of lookups (as a vec of subtables), along with a map of features -> lookups
+    /// (by order in the first vec)
+    fn assign_lookups_to_scripts(
```

Comment:
it would be nice for the function to permalink to the python function or block(s) it mirrors.